File: labs/lab6/C++/Text.cpp
#include "Text.h"
#include <cctype>
#include <numeric>
#include <algorithm>

// --- приватне ---
bool Text::IsVowel(char ch) const {
    const std::string vowels = "aeiouAEIOUаеєиіоуюяАЕЄИІОУЮЯ";
    return vowels.find(ch) != std::string::npos;
}

// --- IVowelWorker ---
size_t Text::CountVowels(const std::string& s) const {
    return std::count_if(s.begin(), s.end(),
        [this](char c){ return IsVowel(c); });
}

double Text::VowelPercentage(const std::string& s) const {
    size_t letters = std::count_if(s.begin(), s.end(), ::isalpha);
    size_t vowels  = CountVowels(s);
    return letters ? 100.0 * vowels / letters : 0.0;
}

// --- керування рядками ---
void Text::AddRow(const Row& r){ rows.push_back(r); }

bool Text::RemoveRow(size_t idx){
    if(idx >= rows.size()) return false;
    rows.erase(rows.begin() + idx);
    return true;
}

void Text::Clear(){ rows.clear(); }

bool Text::RemoveRowsWithSubstring(const std::string& sub){
    auto oldSize = rows.size();
    rows.erase(std::remove_if(rows.begin(), rows.end(),
        [&](const Row& r){ return r.Contains(sub); }),
        rows.end());
    return oldSize != rows.size();
}

// --- статистика ---
double Text::AverageRowLength() const{
    if(rows.empty()) return 0;
    size_t total = std::accumulate(rows.begin(), rows.end(), size_t{},
        [](size_t acc, const Row& r){ return acc + r.Length(); });
    return static_cast<double>(total) / rows.size();
}

double Text::TotalVowelPercentage() const{
    size_t letters = 0, vowels = 0;
    for(const auto& r : rows){
        const std::string& s = r.Get();
        letters += std::count_if(s.begin(), s.end(), ::isalpha);
        vowels  += CountVowels(s);
    }
    return letters ? 100.0 * vowels / letters : 0.0;
}

File: labs/lab6/C#/Text.cs
using System.Text;
namespace TextLib;

public class Text : IVowelWorker
{
    private readonly List<Row> _rows = new();

    private static bool IsVowel(char ch)
    {
        const string vowels = "aeiouAEIOUаеєиіоуюяАЕЄИІОУЮЯ";
        return vowels.Contains(ch);
    }

    // --- IVowelWorker ---
    public int CountVowels(string s) => s.Count(IsVowel);

    public double VowelPercentage(string s)
    {
        int letters = s.Count(char.IsLetter);
        int vowels  = CountVowels(s);
        return letters == 0 ? 0 : 100.0 * vowels / letters;
    }

    // --- керування рядками ---
    public void AddRow(Row r) => _rows.Add(r);
    public bool RemoveRow(int idx)
    {
        if (idx < 0 || idx >= _rows.Count) return false;
        _rows.RemoveAt(idx);
        return true;
    }
    public void Clear() => _rows.Clear();

    public bool RemoveRowsWithSubstring(string sub)
    {
        int before = _rows.Count;
        _rows.RemoveAll(r => r.Contains(sub));
        return before != _rows.Count;
    }

    // --- статистика ---
    public double AverageRowLength() =>
        _rows.Count == 0 ? 0 : _rows.Average(r => r.Length);

    public double TotalVowelPercentage()
    {
        int letters = 0, vowels = 0;
        foreach (var r in _rows)
        {
            letters += r.Data.Count(char.IsLetter);
            vowels  += CountVowels(r.Data);
        }
        return letters == 0 ? 0 : 100.0 * vowels / letters;
    }
}
